via/services/youtube_transcript.py

```diff
@@ -0,0 +1,147 @@
+class TranscriptInfo:
+    """Information about a YouTube video transcript."""
+
+    language_code: str
+    """The transcript's language code, e.g. "en-is"."""
+
+    name: str
+    """The transcript's name, e.g. "English (United States)"."""
+
+    url: str
+    """The transcript's download URL."""
+
+    autogenerated: bool
+    """Whether or not the transcript autogenerated by speech recognition."""
+
+    @property
+    def id(self) -> str:  # pylint:disable=invalid-name
+        """Return a unique ID for this transcript."""
+        return ".".join(
+            [
+                self.language_code,
+                "a" if self.autogenerated else "",
+                b64encode(self.name.encode("utf-8")).decode("utf-8"),
+            ]
+        ).rstrip(".")
+
+
+#: The preferences that we apply when choosing a default transcript for a video.
+DEFAULT_TRANSCRIPT_PREFERENCES = (
+    {
+        "language_code": re.compile("en"),
+        "name": re.compile("English"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en-.*"),
+        "name": re.compile(r"English \(.*\)"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en"),
+        "name": re.compile(".*"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en-.*"),
+        "name": re.compile(".*"),
+        "autogenerated": False,
+    },
+    {
+        "language_code": re.compile("en"),
+        "name": re.compile(".*"),
+        "autogenerated": True,
+    },
+    {
+        "language_code": re.compile("en-"),
+        "name": re.compile(".*"),
+        "autogenerated": True,
+    },
+)
+
+
+class YouTubeTranscriptService:
+    """A service for getting text transcripts of YouTube videos."""
+
+    def __init__(self, http_service: HTTPService):
+        self._http_service = http_service
+
+    def get_transcript_infos(self, video_id: str) -> List[TranscriptInfo]:
+        """Return the list of available transcripts for `video_id`."""
+        response = self._http_service.post(
+            "https://youtubei.googleapis.com/youtubei/v1/player",
+            json={
+                "context": {
+                    "client": {
+                        "hl": "en",
+                        "clientName": "WEB",
+                        "clientVersion": "2.20210721.00.00",
+                    }
+                },
+                "videoId": video_id,
+            },
+        )
+        json = response.json()
+        dicts = json["captions"]["playerCaptionsTracklistRenderer"]["captionTracks"]
+
+        return [
+            TranscriptInfo(
+                language_code=caption_track_dict["languageCode"].lower(),
+                autogenerated=caption_track_dict.get("kind", None) == "asr",
+                name=caption_track_dict["name"]["simpleText"],
+                url=caption_track_dict["baseUrl"],
+            )
+            for caption_track_dict in dicts
+        ]
```

Comment:
The third-party `youtube-transcript-api` library gets a blob of JSON about the video's transcripts by scraping the video's HTML page. We're doing it by calling this deprecated and undocumented legacy V1 API instead. This is no better in terms of reliability or legitimateness, but the code is simpler because we just call a JSON API rather than scraping HTML.